node/src/api/registration.rs

```diff
@@ -1,181 +1,118 @@
-        use nmos_model::version::is_04::V1_0;
-        use nmos_schema::is_04::v1_0_x::{
-            RegistrationapiResourcePostRequest, RegistrationapiResourcePostRequestHealthVariant3,
-        };
-
-        let resource::ReceiverJson::V1_0(receiver_json) = receiver.to_json(&V1_0);
-        let receiver_post_request = RegistrationapiResourcePostRequestHealthVariant3 {
-            data: Some(receiver_json),
-            type_: Some(String::from("receiver")),
-        };
-        let post_request = RegistrationapiResourcePostRequest::Variant3(receiver_post_request);
-
-        client.post(url.clone()).json(&post_request).send().await?;
+        resource: &dyn resource::Registerable,
+        api_version: &APIVersion,
+    ) -> Result<reqwest::Response, Box<dyn std::error::Error>> {
+        let res = Self::post_resource(client, url, resource, api_version).await?;
+
+        if res.status() == StatusCode::OK {
+            warn!(
+                "Resource already present in API deleting: {}",
+                resource.registry_path()
+            );
+
+            let res = Self::delete_resource(client, url, resource).await?;
+
+            if res.status() == StatusCode::OK {
+                return Err(Box::new(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "Failed to register resource after deleting and re-registering",
+                )));
+            } else {
+                return Ok(res);
+            }
+        }
 
-        Ok(())
+        Ok(res)
     }
 
     pub async fn register_resources(
         client: &reqwest::Client,
         model: Arc<Model>,
-        registry: &NmosMdnsRegistry,
+        registry: Arc<Mutex<Option<NmosMdnsRegistry>>>,
+        api_version: &APIVersion,
     ) -> Result<(), Box<dyn std::error::Error>> {
-        let base = &registry.url.join("v1.0/").unwrap();
+        let registry = registry.lock().await.clone().unwrap();
 
-        info!("Attempting to register with {}", base);
+        let base = &registry
+            .url
+            .join(format!("{}/", api_version.to_string()).as_str())
+            .unwrap();
 
-        // Resource endpoint
         let resource_url = &base.join("resource").unwrap();
 
-        // Get node
-        let nodes = model.nodes().await;
-        let node = nodes.iter().next().unwrap().1;
+        info!("Attempting to register with {}", resource_url);
 
         // Register resources in order
-        Self::register_node(client, resource_url, node).await?;
+        debug!("Registering nodes...");
+        for (_, node) in model.nodes().await.iter() {
+            Self::register_resource(client, resource_url, node, api_version).await?;
+        }
+        debug!("Registering devices...");
         for (_, device) in model.devices().await.iter() {
-            Self::register_device(client, resource_url, device).await?;
+            Self::register_resource(client, resource_url, device, api_version).await?;
```

Comment:
You might be able to chain all iterators of the loops together and apply at the end a simple try_for_each.
That would make things more compact but it depends a bit of you can map all types to a shared one.

Reply:
I tried that for a while but couldn't get it to work due to type incongruences, would likely need a little more implementation effort, but i agree that it would be a lot cleaner